Ofast deprecation clarifications

sjoerdmeijer opened the pull request:
Following up on the RFC discussion, this is clarifying that the main purpose and effect of the -Ofast deprecation is to discourage its usage and that everything else is more or less open for discussion, e.g. there is no timeline yet for removal.

AaronBallman commented on clang/docs/CommandGuide/clang.rst:
```diff
@@ -429,8 +429,11 @@ Code Generation Options
-    in combination with :option:`-ffast-math`.
+    language standards. This is deprecated in Clang-19 and a warning is emitted
+    that :option:`-O3` in combination with :option:`-ffast-math` should be used
+    instead if the request for non-standard math behavior is intended. Thus, as
+    there is no timeline yet for removal, the aim is to discourage its usage
+    due to the compliance violating optimizations.
```
```suggestion
    instead if the request for non-standard math behavior is intended. 
    There is no timeline yet for removal; the aim is to discourage use of :option:`-Ofast`
    due to the surprising behavior of an optimization flag changing the observable
    behavior of correct code.
    
```

AaronBallman commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -495,10 +495,12 @@ New Compiler Flags
-  enables the ``-O3`` optimization-level, as well as enabling non-standard
-  ``-ffast-math`` behaviors. As such, it is somewhat misleading as an
-  "optimization level". Users are advised to switch to ``-O3 -ffast-math`` if
+- The ``-Ofast`` command-line option has been deprecated, but there is no
+  timeline for removal yet. Thus, the main effect of emitting a deprecation
+  warning message is to discourage its usage due to the problems of ``-Ofast``:
+  it enables both the ``-O3`` optimization-level as well as non-standard
+  ``-ffast-math`` behaviors and as such it is perceived to be misleading as an
+  optimization level.  Users are advised to switch to ``-O3 -ffast-math`` if
```
```suggestion
- The ``-Ofast`` command-line option has been deprecated, but there is no
  timeline for removal. Deprecation is intended to discourage use of the flag
  due to the problems of ``-Ofast``: it enables both the ``-O3`` optimization
  level as well as non-standard ``-ffast-math`` behaviors and as such it is
  misleading as an optimization level.  Users are advised to switch to
  ``-O3 -ffast-math`` if
```

AaronBallman reviewed (commented):
Thank you for working on this!

I think we need these changes to be against `main` and not `19.x`, aside from the release notes (otherwise we're losing the documentation in 20.x). For the release notes, I think we can talk to @tstellar and @tru about getting the 19.x notes updated once we've finalized the wording for it.

One other change we should probably make at the same time is to update `Options.td` to use `HelpTextForVariants` so the help text for Flang does not suggest the option is deprecated in their driver.

sjoerdmeijer commented:
> I think we need these changes to be against main and not 19.x, aside from the release notes (otherwise we're losing the documentation in 20.x). 

Oh right, yes, missed that, makes sense. Will split this up, let's do the documentation update here first as suggested.

sjoerdmeijer committed:
Ofast documentation deprecation clarifications

Following up on the RFC discussion, this is clarifying that the main purpose
and effect of the -Ofast deprecation is to discourage its usage and that
everything else is more or less open for discussion, e.g. there is no timeline
yet for removal.

AaronBallman reviewed (commented):
FWIW, I'm happy enough with the proposed wording (I want to give others a chance to weigh in before approving though).

sjoerdmeijer committed:
Update clang/docs/CommandGuide/clang.rst

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

Endilll commented on clang/docs/CommandGuide/clang.rst:
```diff
@@ -429,8 +429,12 @@ Code Generation Options
-    in combination with :option:`-ffast-math`.
+    language standards. This is deprecated in Clang 19 and a warning is emitted
+    that :option:`-O3` in combination with :option:`-ffast-math` should be used
+    instead if the request for non-standard math behavior is intended.  There
```
```suggestion
    instead if the request for non-standard math behavior is intended. There
```

sjoerdmeijer committed:
Update clang/docs/CommandGuide/clang.rst

Co-authored-by: Vlad Serebrennikov <serebrennikov.vladislav@gmail.com>

jyknight reviewed (approved):
LGTM, thanks.

sjoerdmeijer commented:
@AaronBallman, shall we merge this?

AaronBallman reviewed (approved):
LGTM, let's go ahead and merge this and get it backported to 19.x, thank you!

sjoerdmeijer commented:
> LGTM, let's go ahead and merge this and get it backported to 19.x, thank you!

Thanks for your help with this.

I am mostly unfamiliar with the release process. Do you know how we can achieve this? Do we suggest this commit to the release manager for a backport and cherry pick? Or do we prepare a patch?

We could add similar clarifications to the release notes, which were initially included in an earlier version of this patch. So same question how we would need to approach that.

AaronBallman commented:
We have instructions at https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches but the basic gist of it is to add a special comment and add the PR to the correct milestone. As in:

/cherry-pick 48d4d4b641702bf6db03a1bac73b7e13dea28349

AaronBallman commented:
> We could add similar clarifications to the release notes, which were initially included in an earlier version of this patch. So same question how we would need to approach that.

In terms of updating the release notes, once the cherry-pick happens, there's an automated bot message asking if you want to add a release note for the changes. I think the best way forward would be to follow those instructions and ask the release managers to help apply the changes.
